src/internet_identity_interface/src/lib.rs

```diff
@@ -173,3 +173,92 @@ pub struct HttpResponse {
 pub struct InternetIdentityInit {
     pub assigned_user_number_range: (UserNumber, UserNumber),
 }
+
+// Archive specific types
+
+#[derive(Eq, PartialEq, Clone, Debug, CandidType, Deserialize)]
+pub enum OperationType {
+    #[serde(rename = "register_anchor")]
+    RegisterAnchor {
+        initial_device: DeviceDataWithoutAlias,
+    },
+    #[serde(rename = "add_device")]
+    AddDevice { new_device: DeviceDataWithoutAlias },
+    #[serde(rename = "update_device")]
+    UpdateDevice {
+        updated_device: PublicKey,
+        changed_data: DeviceDataUpdate,
+    },
+    #[serde(rename = "remove_device")]
+    RemoveDevice { removed_device: PublicKey },
+}
+
+#[derive(Eq, PartialEq, Clone, Debug, CandidType, Deserialize)]
+pub struct LogEntry {
+    // store user_number in LogEntry, such that anchor operations can be attributed to a user without consulting the index.
+    pub user_number: UserNumber,
+    pub operation: OperationType,
+    pub timestamp: Timestamp,
+    pub caller: Principal,
+    // global sequence number to detect lost messages (if any)
+    pub sequence_number: u64,
+}
+
+#[derive(Eq, PartialEq, Clone, Debug, CandidType, Deserialize)]
+pub struct DeviceDataWithoutAlias {
+    pub pubkey: DeviceKey,
+    pub credential_id: Option<CredentialId>,
+    pub purpose: Purpose,
+    pub key_type: KeyType,
+    pub protection: DeviceProtection,
+}
+
+// If present, the attribute has been changed to the value given.
+// Does not include the pubkey because it cannot be changed.
+#[derive(Eq, PartialEq, Clone, Debug, CandidType, Deserialize)]
+pub struct DeviceDataUpdate {
+    pub alias: Option<Hidden>,
+    pub credential_id: Option<CredentialId>,
+    pub purpose: Option<Purpose>,
+    pub key_type: Option<KeyType>,
+    pub protection: Option<DeviceProtection>,
+}
+
+// Placeholder for information that has been hidden for privacy reasons.
+#[derive(Eq, PartialEq, Clone, Debug, CandidType, Deserialize)]
+pub enum Hidden {
+    HiddenForPrivacyReasons,
+}
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+pub struct Logs {
+    // make this a vec of options to keep LogEntry extensible
+    pub entries: Vec<Option<LogEntry>>,
+    // index pointing to the next entry not included in this response, if any
+    pub next_idx: Option<u64>,
+}
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+pub struct UserLogs {
+    // make this a vec of options to keep LogEntry extensible
+    pub entries: Vec<Option<LogEntry>>,
+    // cursor pointing to the next entry not included in this response, if any
+    pub cursor: Option<Cursor>,
+}
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+pub enum Cursor {
+    // timestamp of the next entry not included in this response, if any
+    #[serde(rename = "timestamp")]
+    Timestamp { timestamp: Timestamp },
+    // index of the next entry not included in this response, if any
+    #[serde(rename = "next_token")]
+    NextToken { next_token: ByteBuf },
```

Comment:
out of curiosity, why "token"? also should the `next_token` be an opaque type like `Index` or something, or does it need to be a `ByteBuf` (because somehow read and interpreted by the client)?

Reply:
I took the `token` terminology from the Twitter pagination api.

Yes the `next_token` should be treated as an opaque blob (which `ByteBuf` represents). It needs correct typings (i.e. instead of candid `reserved`) because the client needs to decode it as part of the response an re-encode it as part of the next request.